// @aws-cdk/aws-cognito-identitypool-alpha/lib/identitypool-user-pool-authentication-provider.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserPoolAuthenticationProvider = void 0;
const jsiiDeprecationWarnings = require("../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const core_1 = require("aws-cdk-lib/core");
const constructs_1 = require("constructs");
/**
 * Defines a User Pool Authentication Provider
 */
class UserPoolAuthenticationProvider {
    constructor(props) {
        try {
            jsiiDeprecationWarnings._aws_cdk_aws_cognito_identitypool_alpha_UserPoolAuthenticationProviderProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, UserPoolAuthenticationProvider);
            }
            throw error;
        }
        this.userPool = props.userPool;
        this.userPoolClient = props.userPoolClient || this.userPool.addClient('UserPoolAuthenticationProviderClient');
        this.disableServerSideTokenCheck = props.disableServerSideTokenCheck ?? false;
    }
    bind(scope, identityPool, _options) {
        try {
            jsiiDeprecationWarnings._aws_cdk_aws_cognito_identitypool_alpha_IIdentityPool(identityPool);
            jsiiDeprecationWarnings._aws_cdk_aws_cognito_identitypool_alpha_UserPoolAuthenticationProviderBindOptions(_options);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.bind);
            }
            throw error;
        }
        constructs_1.Node.of(identityPool).addDependency(this.userPool);
        constructs_1.Node.of(identityPool).addDependency(this.userPoolClient);
        const region = core_1.Stack.of(scope).region;
        const urlSuffix = core_1.Stack.of(scope).urlSuffix;
        return {
            clientId: this.userPoolClient.userPoolClientId,
            providerName: `cognito-idp.${region}.${urlSuffix}/${this.userPool.userPoolId}`,
            serverSideTokenCheck: !this.disableServerSideTokenCheck,
        };
    }
}
exports.UserPoolAuthenticationProvider = UserPoolAuthenticationProvider;
_a = JSII_RTTI_SYMBOL_1;
UserPoolAuthenticationProvider[_a] = { fqn: "@aws-cdk/aws-cognito-identitypool-alpha.UserPoolAuthenticationProvider", version: "2.116.1-alpha.0" };
